### PR TITLE
onStructureGrow()

### PR DESCRIPTION
**Pull request by md-5**
@dinnerbone
@oliverw92
@diddiz
@iffa

You have all heard enough of this, but I present you the onStructureGrow() event, for all your mushroom and tree growth logging needs, complete with bonemeal, player etc.

Please pull ASAP.

Craftbukkit: https://github.com/Bukkit/CraftBukkit/pull/533

### CODE REVIEW

**Comment by stephenwilliams** [comment]
They will probably ask you to put it all in one commit

**Comment by md-5** [comment]
Yay I squashed it!

**Comment by stephenwilliams** [comment]
:)

**Comment by md-5** [comment]
I did it by making a new branch, resetting to before
My commits and then merging with a squash commit
Is that the correct way?

**Comment by DiddiZ** [comment]
yay, pls pull

**Comment by Zidane** [comment]
Make a note in your comments for JavaPluginLoader that its for mushrooms as well (for consistency)

Regardless of that...this will probably be yet again another pull in which is useful, needed, and ultimately ignored and forgotten.

**Comment by Zidane** [comment]
Also under StructureGrowEvent, I see a TreeType but shouldn't that be a StructureType?  As TreeType makes no sense in regards to mushrooms (as they aren't a tree)

....sorry for the double post.

**Comment by md-5** [comment]
pluginloader, sure, however regarding treetype enum, we can't rename it incase any plugins rely on it, and creating a duplicate enum with only two different fields is just a waste of space. To me a mushroom the size of a tree, with a trunk and 'leaves' is a tree.

**Comment by DrakiaXYZ** [comment]
@NinjaZidane
"this will probably be yet again another pull in which is useful, needed, and ultimately ignored and forgotten."
Too true, it's a shame how badly this project is run. Here's hoping somebody makes a fork and actually accepts community input as opposed to ignoring, silencing, and censoring it :)

**Comment by md-5** [comment]
We need this soo badly for Guardian, tree grief is the new grief.
There will be a new pull system soon I hear.

**Comment by feildmaster** [comment]
@md-5 : More like.. a new bleed system where anyone can make anything, and if it meets their requirements they will put it in. 

@TheDgtl : They run the community fine, get off your high horse and stop blaming others for your incompetence.

Also, on a completely unrelated note, You shouldn't be afraid of breaking plugins. Break them. It's what [plugin] updates are for.

**Comment by md-5** [comment]
Well it already breaks Essentials and its bad use of treegen code. I know the bleed system, I was just making a generic statement.

**Comment by DrakiaXYZ** [comment]
@feildmaster
Yes, because I'm the only one who thinks that. Oh wait, no I'm not :)
I take the blame for my own wrongdoings and issues. Sadly the Bukkit team does not do the same.

**Comment by md-5** [comment]
I would like to comment, but I'm not :)

**Comment by DrakiaXYZ** [comment]
Sorry for arguing in your PR @md-5 :)
I agree that this needs to be pulled, or the Bukkit team needs to get in gear and implement something themselves. I've had to deal with tree griefing, and it's a royal pain in the behind. So here's a +1 from the banished Drakia :)

**Comment by md-5** [comment]
I wasn't arguing with you :)
But all the same I'm not going to state my opinion.

**Comment by iffa** [comment]
Javadocs are a bit off here still. :DD

**Comment by md-5** [comment]
Comment on the individual lines please

File: src/main/java/org/bukkit/event/Event.java [hunk 1]
```diff
@@ -587,6 +587,12 @@ public abstract class Event implements Serializable {
          * @see org.bukkit.event.world.PortalCreateEvent
          */
         PORTAL_CREATE (Category.WORLD),
+        /**
+         * Called when a tree attempts to grow (sapling -> tree)
```
Comment (iffa) [inline src/main/java/org/bukkit/event/Event.java:591]: Javadocs

File: src/main/java/org/bukkit/event/Event.java [hunk 2]
```diff
@@ -587,6 +587,12 @@ public abstract class Event implements Serializable {
...
+         * 
+         * @see org.bukkit.event.world.TreeGrowEvent
```
Comment (iffa) [inline src/main/java/org/bukkit/event/Event.java:593]: StructureGrowEvent

File: src/main/java/org/bukkit/event/world/StructureGrowEvent.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        super(Type.STRUCTURE_GROW, location.getWorld());
+        this.location = location;
+        this.species = species;
+        this.bonemeal = bonemeal;
+        this.player = player;
+        this.blocks = blocks;
+    }
+
+    /**
+     * Gets the location of the tree.
```
Comment (iffa) [inline src/main/java/org/bukkit/event/world/StructureGrowEvent.java:32]: structure

File: src/main/java/org/bukkit/event/world/StructureGrowEvent.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+    /**
+     * Gets the player that bonemealed the structure.
+     * 
+     * @return Player that bonemealed the structure, null if was not bonemealed
+     */
+    public Player getPlayer() {
+        return player;
+    }
+    public ArrayList<BlockState> getBlocks() {
```
Comment (iffa) [inline src/main/java/org/bukkit/event/world/StructureGrowEvent.java:66]: missing entirely

File: src/main/java/org/bukkit/event/world/WorldListener.java [hunk 5]
```diff
@@ -45,6 +45,13 @@ public class WorldListener implements Listener {
     public void onPortalCreate(PortalCreateEvent event) {}
 
     /**
+     * Called when a tree is grown (sapling -> full tree), natural or not (bonemeal)
```
Comment (iffa) [inline src/main/java/org/bukkit/event/world/WorldListener.java:48]: structure is grown

**Comment by md-5** [comment]
@iffa
I'll add the JavaDocs for 1.0
@dinnerbone Is the best way to handle the 1.0 changes do a fresh fork and push? We really need this.

**Comment by don4of4** [comment]
Lets get this done and pulled.  This is really a critical fix.

**Comment by md-5** [comment]
@don4of4
Unfortuantely I have to either wait for mc-dev to become public or decompile it this weekend :(
I want to get it done asap.
Hopefully bleeding into Bukkit comes soon!

**Comment by iffa** [comment]
Where do you need mc-dev for onStructureGrow()?

**Comment by md-5** [comment]
@iffa
To get the latest decompiled sources for 1.0, so I can update it to make sure it is 100% ready to be pulled. If its not out this weekend I'll use my decompiled sources. A 10 seconds look reveals the decompilation accuracy is very good.

**Comment by CruzBishop** [comment]
@md-5 beta 1.9 isn't going to be released, 1.0(.0) was released at minecon.

**Comment by md-5** [comment]
@cruzbishop
Me iz no idiotz
I can haz knows what mez talks abouts.
Twas a sympl tyypo.

**Comment by CruzBishop** [comment]
NOTCH WE WANT 1.9, YOU LAZY SWEDE :c

Yeah, I guessed, just thought I should point it out :P

**Comment by md-5** [comment]
https://github.com/Bukkit/Bukkit/pull/425

1.0.0 Version

**Pull request closed without merging**